core/base/continuousScatterPlot/ContinuousScatterPlot.h

```diff
@@ -233,53 +232,86 @@ int ttk::ContinuousScatterPlot::execute(
+    bool isInTriangle{}; // True if the tetra is class 0
+
+    bool zCrossProductsSigns[4]
+      = {(data[1][0] - data[0][0]) * (data[2][1] - data[0][1])
+             - (data[1][1] - data[0][1]) * (data[2][0] - data[0][0])
+           > 0,
+         (data[2][0] - data[1][0]) * (data[3][1] - data[1][1])
+             - (data[2][1] - data[1][1]) * (data[3][0] - data[1][0])
+           > 0,
+         (data[3][0] - data[2][0]) * (data[0][1] - data[2][1])
+             - (data[3][1] - data[2][1]) * (data[0][0] - data[2][0])
+           > 0,
+         (data[0][0] - data[3][0]) * (data[1][1] - data[3][1])
+             - (data[0][1] - data[3][1]) * (data[1][0] - data[3][0])
+           > 0};
+
+    // For class 0, the quad is not convex, which means
+    // all but one consecutive edge vector cross-product Z coordinate have the
+    // same sign.
+    if((zCrossProductsSigns[0] != zCrossProductsSigns[1])
+       != (zCrossProductsSigns[2] != zCrossProductsSigns[3])) {
       isInTriangle = true;
-      index[0] = 0;
-      index[1] = 1;
-      index[2] = 3;
-      index[3] = 2;
-    } else if(Geometry::isPointInTriangle(data[0], data[2], data[3], data[1])) {
-      isInTriangle = true;
-      index[0] = 0;
-      index[1] = 2;
-      index[2] = 3;
-      index[3] = 1;
-    } else if(Geometry::isPointInTriangle(data[1], data[2], data[3], data[0])) {
-      isInTriangle = true;
-      index[0] = 1;
-      index[1] = 2;
-      index[2] = 3;
-      index[3] = 0;
+      if(zCrossProductsSigns[1] == zCrossProductsSigns[2]
+         && zCrossProductsSigns[2] == zCrossProductsSigns[3]) {
+        index[0] = 0;
+        index[1] = 2;
+        index[2] = 3;
+        index[3] = 1;
+      } else if(zCrossProductsSigns[0] == zCrossProductsSigns[2]
+                && zCrossProductsSigns[2] == zCrossProductsSigns[3]) {
+        index[0] = 0;
+        index[1] = 1;
+        index[2] = 3;
+        index[3] = 2;
+      } else if(zCrossProductsSigns[0] == zCrossProductsSigns[1]
+                && zCrossProductsSigns[1] == zCrossProductsSigns[2]) {
+        index[0] = 1;
+        index[1] = 2;
+        index[2] = 3;
+        index[3] = 0;
+      }
     }
 
     // projection:
     double density{};
-    double imaginaryPosition[3]{};
+    double imaginaryPosition[3]{0, 0, 0};
     triangles.clear();
     std::array<SimplexId, 3> triangle{};
-    // class 0
+
+    // class 0 projection : 3 triangles
     if(isInTriangle) {
       // mass density
       double massDensity{};
       {
-        double A;
+        double fullArea;
 
         Geometry::computeTriangleArea(
-          data[index[0]], data[index[1]], data[index[2]], A);
-        double invA = 1.0 / A;
-        if(A == 0.) {
-          invA = 0.0;
+          data[index[0]], data[index[1]], data[index[2]], fullArea);
+
+        double invArea;
+        if(fullArea == 0.) {
+          invArea = 0.0;
           isLimit = true;
         }
+        invArea = 1.0 / fullArea;
```

Comment:
I believe you want to put an `else` statement before this line (to avoid divisions by zero).

Reply:
fixed